File: leetcode-in-python3/solution/11_container_with_most_water.py
from typing import List

class Solution:
    def maxArea(self, height: List[int]) -> int:
        ans = 0
        l, r = 0, len(height) - 1

        while l < r:
            if height[l] < height[r]:
                area = height[l] * (r - l)
                l += 1
            else:
                area = height[r] * (r - l)
                r -= 1
            ans = max(ans , area)
        return ans


import unittest


class TestSolution(unittest.TestCase):

    def setUp(self):
        self.input = [
            [1, 8, 6, 2, 5, 4, 8, 3, 7],
            [1, 1],
            [4, 3, 2, 1, 4]
        ]
        self.answer = [49, 1, 16]
        self.s = Solution()

    def test_solution(self):
        for i, item in enumerate(self.input):
            ans = self.s.maxArea(item)
            self.assertEqual(ans , self.answer[i],ans )


if __name__ == '__main__':
    unittest.main()

File: leetcode-in-rust/src/solution/_11_container_with_most_water.rs
pub struct Solution {}

impl Solution {
    pub fn max_area(height: Vec<i32>) -> i32 {
        let (mut l, mut r) = (0usize, height.len() - 1);
        let mut ans = 0;
        let mut area;
        while l < r {
            if height[l] < height[r] {
                area = height[l] * (r - l) as i32;
                l += 1
            } else {
                area = height[r] * (r - l) as i32;
                r -= 1
            }
            if ans < area {
                ans = area
            }
        }
        ans
    }
}


#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_solution() {
        let test_case = vec![
            (vec![1, 8, 6, 2, 5, 4, 8, 3, 7], 49),
            (vec![1, 1], 1),
            (vec![4, 3, 2, 1, 4], 16),
            (vec![1, 2, 1], 2)
        ];
        for (nums, answer) in test_case {
            let ans = Solution::max_area(nums);

            assert_eq!(ans, answer);
        }
    }
}
